```{r} 
#| label: a1
#| echo: false

reactable <- function(...) {
  reactable::reactable(..., wrap = F, resizable = T, bordered = T)
}

knit_print.data.frame = function(x, ...) {
  html = as.character(htmltools::as.tags(reactable(x,static = TRUE)))
  res = paste0("<div class='kable_wrapper'>", html, "</div>")
  knitr::asis_output(res)
}

registerS3method(
  "knit_print", "data.frame", knit_print.data.frame,
  envir = asNamespace("knitr")
)

```

```{r}
#| label: a2
#| echo: false
library(tidyverse)
faculty <- read_csv("data/faculty.csv")
courses <- read_csv("data/courses.csv")
genes <- readRDS("data/rowdata.rds")
clin <- readRDS("data/coldata.rds")
assay <- readRDS("data/assay.rds")
```

## Attribution
These slides draw heavily from slides created by [@meghall06].

## Outline

<br>

::: {.incremental}
- What is R? 
- Basic data manipulation in R
- Using R effectively
- More data manipulation
- Visualizing data
- Bioconductor
- The analysis of expression data
- A peek at advanced topics
:::

# What is R?

::: {.numbers}
1 [2 3 4 5 6]{style="opacity:0.25"}
:::

---

## What is R?

::: {style="text-align: center"}
R is an **open-source** scripting language and environment for working with data.

Open-source means the software is free of charge. most importantly it means the we are **free** to inspect, use, modify and redistribute the software.
:::

---

## What is R?

The R language is the vehicle of choice for research in statistical methodology, bioinformatics research and pharmaceutical industry.

---

## Why Not Excel?

::: {.incremental}
- Limited Data Handling

- Reproducibility Challenges

- Basic Statistical Tools

- Visualization Limitations

- Prone to Human Errors

- Lack of Automation

- No Version Control

- Insufficient Integration

- Scalability Constraints

- Security Concerns
:::

:::{.notes}
- **Limited Data Handling:** Explain scenarios where Excel cannot manage the size of biomedical datasets.
- **Reproducibility Challenges:** Discuss the importance of reproducible research and how Excel’s manual processes hinder this.
- **Basic Statistical Tools:** Highlight the need for advanced statistical analyses in biomedical research that Excel cannot provide.
- **Visualization Limitations:** Show examples of complex visualizations needed in research that are difficult to create in Excel.
- **Prone to Human Errors:** Share instances where manual errors in Excel could lead to significant research inaccuracies.
- **Lack of Automation:** Emphasize the time lost on repetitive tasks in Excel and the benefits of automation.
- **Version Control Issues:** Illustrate the difficulties in tracking changes and collaborating using Excel.
- **Insufficient Integration:** Explain the necessity of integrating with other research tools and databases, programming languages
- **Scalability Constraints:** Discuss how research projects grow and Excel's limitations in scaling with them.
- **Security Concerns:** Address the importance of data security in biomedical research and Excel’s shortcomings.
:::

---

## Why R?

::: {.incremental}
- Reproducible Analysis

- Handles Large & Complex Datasets

- Advanced Statistical Tools

- Superior Data Visualization

- Data and code are separate

- Automation & Efficiency

- Extensive Package Ecosystem

- Integration Capabilities

- Scalability & Performance
:::

:::{.notes}
- **Reproducible Analysis:** Demonstrate how R scripts can be reused and shared to ensure consistent results.
- **Handles Large & Complex Datasets:** Compare R’s capacity to handle large datasets versus Excel’s limitations.
- **Advanced Statistical Tools:** Provide examples of advanced statistical techniques available in R that are essential for biomedical research.
- **Superior Data Visualization:** Showcase some high-quality visualizations created with `ggplot2` to illustrate R’s capabilities.
- **Automation & Efficiency:** Explain how R can automate data processing tasks, reducing manual effort and errors.
- **Extensive Package Ecosystem:** Highlight key R packages that are particularly useful in biomedical sciences.
- **Active Community Support:** Mention resources like CRAN, Bioconductor, and community forums that support R users.
- **Integration Capabilities:** Describe how R can interface with databases, other programming languages, and research tools.
- **Scalability & Performance:** Discuss R’s ability to handle increasing data sizes and more complex analyses efficiently.
- **Enhanced Security:** Talk about R’s options for securing data and managing access, which are critical in biomedical research.
:::

---

## Getting R

::: {.logos}
[![](https://www.r-project.org/logo/Rlogo.svg)](https://cloud.r-project.org/)
[![](https://upload.wikimedia.org/wikipedia/commons/d/d0/RStudio_logo_flat.svg)](https://posit.co/download/rstudio-desktop/#download)

or: 

[![](http://mybinder.org/badge_logo.svg)](http://mybinder.org/v2/gh/she3o/whyR/master?urlpath=rstudio)
:::


---

## Navigating RStudio

![](images/RStudio1.webp)

. . .

::: {.absolute top="50%" left="10%"}
::: {style="color: #cc0000"}
code *can* go here
:::
:::

---

## Navigating RStudio

![](images/RStudio2.webp)


::: {.absolute top="35%" left="25%"}
::: {style="color: #cc0000"}
code can also<br>go here
:::
:::

---

## Using R

::: columns
::: {.column width="50%"}
<br>
<br>
You use R via *packages*
:::
::: {.fragment}
<br/>
...which contain *functions*
:::
::: {.fragment}
<br>
...which are just *verbs*
:::
:::

::: {.column width="50%"}
::: {.fragment}
![](images/tidyverse.webp){.absolute top="70" left="530" height="574" width="499"}
:::
:::

---

## data


::: {style="text-align: center"}
`genes` has `{r} ncol(genes) ` columns and `{r} nrow(genes)` rows.
:::

<br>

::: {style="font-size: 0.75em"}
```{r}
#| label: a3
#| echo: false
genes
```
:::

---

## data

::: {style="text-align: center"}
`clin` has `{r} ncol(clin) ` columns and `{r} nrow(clin)` rows.
:::

::: {style="font-size: 0.75em"}
```{r}
#| label: a4
#| echo: false
clin
```
:::

## data

::: {style="text-align: center"}
`assay` is a `{r} ncol(assay) `x`{r} nrow(assay)` matrix.
:::

<br>

```{r}
#| label: a5
#| echo: false
display_assay <- assay[1:100,1:100]
colnames(display_assay) <- NULL
rownames(display_assay) <- NULL
display_assay
```

# Basic data manipulation

::: {.numbers}
[1]{style="opacity:0.25"} 2 [3 4 5 6]{style="opacity:0.25"}
:::

---

## Useful operators

<br>
<br>

::: columns
::: {.column width="15%"}
::: {style="font-size: 1.75em; text-align: center"}
`<-` 
:::
:::

::: {.column width="30%"}
::: {style="font-size: 1.5em; text-align: center"}
"save as" 
:::
:::
:::

. . .

::: columns
::: {.column width="15%"}
::: {style="font-size: 1.75em; text-align: center"}
`|>` 
:::
:::

::: {.column width="30%"}
::: {style="font-size: 1.5em; text-align: center"}
"and then" 
:::
:::
:::

---

## Common functions

`filter` keeps or discards rows (aka observations)

`select` keeps or discards columns (aka variables)

`arrange` sorts data set by certain variable(s)

`count` tallies data set by certain variable(s)

`mutate` creates new variables

`group_by`/`summarize` aggregates data ([pivot tables]{style="color:#cc0000"}!)

`str_*` functions work easily with text

---

## Syntax of a function


`function(data, argument(s))`


. . .

is the same as


`
data |>`
  `function(arguments(3)) `
---

## Filter


`filter` keeps or discards rows (aka observations)

the `==` operator tests for equality

. . .

```{r}
#| label: a6
genes |>
  filter(gene_type == "miRNA")
```

---

## Filter

the `|` operator signifies "or"

. . .

```{r}
#| label: a7
clin |> 
  filter(
    paper_pathologic_stage == "Stage_II" |
    paper_pathologic_stage == "Stage_III"
  )
```

---

## Filter

the `%in%` operator allows for multiple options in a list

. . .

```{r}
#| label: a8
genes |> 
  filter(
    gene_name %in% c("BRCA1", "TP53", "PTEN")
  )
```

---

## Filter

the `&` operator combines conditions

```{r}
#| label: a9
clin |> 
  filter(
    paper_pathologic_stage %in% c("Stage_I", "Stage_II") &
    year_of_birth > 1970
  )
```

---

## Select

`select` keeps or discards columns (aka variables)

```{r}
#| label: a10
clin |> 
  select(
    patient,
    ajcc_pathologic_n,
    ajcc_pathologic_m,
    ajcc_pathologic_t
  )
```

---

## Select

can drop columns with `- column`

```{r}
#| label: a11
clin |> 
  select(
    -barcode,
    -sample,
    -sample_submitter_id
  )
```

---

## Select

the pipe `|>` chains multiple functions together

```{r}
#| label: a12
genes |> 
  filter(
    gene_type == "lncRNA"
  ) |>
  pull(gene_name)
```

---

## Arrange

`arrange` sorts data set by certain variable(s)

use `desc()` to get descending order

```{r}
#| label: a13
clin |> 
  arrange(desc(age_at_diagnosis))
```

---

## Count

`count` tallies data set by certain variable(s) (very useful for familiarizing yourself with data)

```{r}
#| label: a14
clin |> 
  count(gender)
```

---

## Count

can use `sort = TRUE` to order results

```{r}
#| label: a15
clin |> 
  count(paper_pathologic_stage, sort = TRUE)
```

---

## Count

You can tally across all combinations of multiple variables

```{r}
#| label: a16
clin |> 
  count(
    paper_pathologic_stage,
    paper_BRCA_Pathology,
    sort = TRUE
  )
```

---

## Mutate

`mutate` creates new variables (with a single `=`)

```{r}
#| label: a17
clin |> 
  mutate(source = "TCGA")
```

---

## Mutate

much more useful with a conditional such as `ifelse()`, which has three arguments:

condition, value if true, value if false

```{r}
#| label: a18
clin |> count(ajcc_pathologic_m)
```
---

```{r}
#| label: a18
clin |> 
  mutate(
    metastatic = ifelse(
      ajcc_pathologic_m %in% c("M0", "cM0 (i+)") ,
      0, 1
    )
  ) |>
  select(metastatic, ajcc_pathologic_m)
```

---

## Mutate

the `!` operator means *not* 

`is.na()` identifies null values

`case_when` generalize `ifelse` to more than two categories

```{r}
#| label: a19
clin |> 
  mutate(
    metastatic = ifelse(
      !is.na(ajcc_pathologic_m) &
      ajcc_pathologic_m %in% c("M0", "cM0 (i+)") ,
      0,1
    )
  ) |>
  select(metastatic, ajcc_pathologic_m)
```

---

## Mutate

with multiple conditions, `case_when()` is much easier!

```{r}
#| label: a20
clin |> 
  mutate(
    t = case_when(
      ajcc_pathologic_t %in% c("T1", "T1a", "T1b", "T1c") ~ "T1",
      ajcc_pathologic_t %in% c("T2", "T2a", "T2b") ~ "T2",
      ajcc_pathologic_t %in% c("T3", "T3a") ~ "T3",
      ajcc_pathologic_t %in% c("T4", "T4b", "T4d") ~ "T4",
      ajcc_pathologic_t == "TX" ~ NA,
      TRUE ~ ajcc_pathologic_t
    )
  ) |>
  select(t, ajcc_pathologic_t)
```

---

## Mutate

Here is real-life code

```{r}
#| label: a21
clin |> 
  mutate(
    t = case_when(
      str_starts(ajcc_pathologic_t, "T1") ~ "T1",
      str_starts(ajcc_pathologic_t, "T2") ~ "T2",
      str_starts(ajcc_pathologic_t, "T3") ~ "T3",
      str_starts(ajcc_pathologic_t, "T4") ~ "T4",
      ajcc_pathologic_t == "TX" ~ NA,
      TRUE ~ ajcc_pathologic_t
    )
  ) |>
  select(t, ajcc_pathologic_t)
```
---

## Mutate

Here is real-life code

```{r}
#| label: a22
genes <- genes %>%
  mutate(
    full_gene_name = AnnotationDbi::mapIds(
      org.Hs.eg.db::org.Hs.eg.db,
      keys = gene_name,
      column = "GENENAME",
      keytype = "SYMBOL",
      multiVals = "first"
    )
  )

  genes |> select(
    gene_name,
    full_gene_name
  )
```

---

## Group by / summarize

`group_by`/`summarize` aggregates data (**pivot tables**!)

`group_by()` identifies the grouping variable(s) and `summarize()` specifies the aggregation 

. . .

```{r}
#| label: a23
courses |> 
 group_by(dept, semester) |>
  summarize(
    enr = sum(enrollment)
  )
```

---

## Group by / summarize

useful arguments within `summarize`: 

`mean`, `median`, `sd`, `min`, `max`, `n`

```{r}
#| label: a24
courses |> 
  group_by(dept, semester) |>
  summarize(
    enr = sum(enrollment),
    count = n_distinct(course_id)
  )
```

# Using R effectively

::: {.numbers}
[1 2]{style="opacity:0.25"} 3 [4 5 6]{style="opacity:0.25"}
:::

---

## Working in RStudio

::: columns
::: {.column width="50%"}
**Typing in the console**

- think of it like a post-it: useful for quick notes but disposable

- actions are saved but code is not

- one chunk of code is run at a time (`Return`)
:::

::: {.column width="50%"}
::: {.fragment}
**Typing in a code file**

- script files have a `.R` extension

- code is saved and sections of any size can be run (`Cmd + Return`)

- do ~95% of your typing in a code file instead of the console!
:::
:::
:::

---

## Working with packages

packages need to be installed on each computer you use

```{r}
#| label: a25
#| eval: false
install.packages("tidyverse") # <1>
```
1. only need to do this once (per computer)

. . .

Most commonly validated packages exist on CRAN.
So they can be installed with `install.packages`.

. . .

You can browse all CRAN packages [here](https://cran.r-project.org/web/views/)

---

## Working with packages

Packages that specialize in the analysis biological data are often hosted on **Bioconductor**.

. . .

They can be installed as follows:

```{r} 
#| label: a26
#| eval: false
if (!require("BiocManager", quietly = TRUE)) # <1>
    install.packages("BiocManager") # <1>

BiocManager::install("limma") # <2>
```
1. Install the Bioconductor installer like any other package
2. Install packages hosted on Bioconductor

. . .

You can browse all R Bioconductor packages [here](https://bioconductor.org/packages/release/BiocViews.html)

---

## Working with packages

[R-universe](https://r-universe.dev/search) indexes all R packages on the web.

. . .

Journals and regulatory bodies don't allow the use of arbitrary packages in submitted documents!

---

## Working with packages

<br>

packages need to be loaded/attached with `library()` at the beginning of every session

```{r}
#| label: a27
#| eval: false
# always put the necessary packages at the top of a code file
library(tidyverse)
```

<br>

::: {.fragment}
can access help files by typing `??tidyverse` or `??mutate` in the console
:::

---

## Organizing with projects

<br>

[**highly**]{style="color: #cc0000"} recommend using *projects* to stay organized

<br>

keeps code files and data files together, allowing for easier file path navigation and better reproducible work habits

. . .

<br>

`File -> New Project`

<br>

more guidance: [here](https://support.posit.co/hc/en-us/articles/200526207-Using-RStudio-Projects) and [here](https://r4ds.had.co.nz/workflow-projects.html#rstudio-projects)

---

## Accessing data

use `read_csv()` to import a csv file

```{r}
#| label: a28
#| eval: false
# the file path is this simple if you use projects!
# ?read_csv() in the console will bring up the help file with more options
faculty <- read_csv("faculty.csv")
```

---

## Accessing data

the [readxl](https://readxl.tidyverse.org/) package is helpful for Excel files

. . .


```{r}
#| label: a29
#| eval: false
# needs to be loaded but not installed as it's part of the tidyverse
library(readxl)
faculty <- read_excel("faculty.xlsx", sheet = 2)
```

. . .

view the data with `View(faculty)` or by clicking on the data name in the Environment pane

---

## Accessing TCGA data

Most biological databases offer R packages that automate data retrieval.

```{r} 
#| label: a30
#| eval: false
library(TCGAbiolinks) # <1>

query <- GDCquery( # <2>
  project = "TCGA-BRCA", # <2>
  data.category = "Transcriptome Profiling", # <2>
  data.type = "Gene Expression Quantification", # <2>
  workflow.type = "STAR - Counts" # <2>
) # <2>

GDCdownload(query) # <3>
assay <- GDCprepare(query) # <4>
```
1. Load the `TCGAbiolinks` package to interact with TCGA servers.
2. Send a query to TCGA.
3. Download the queried data to your computer.
4. Transform and import the downloaded data to a suitable format.

# More data manipulation

::: {.numbers}
[1 2 3]{style="opacity:0.25"} 4 [5 6]{style="opacity:0.25"}
:::

---

## Stringr functions

functions from `stringr` (which all start with `str_`) are useful for working with text data

```{r}
#| label: a31
faculty |> 
  filter(str_detect(rank, "Professor"))
```

---

## Stringr functions

cheat sheet of functions is [here](https://github.com/rstudio/cheatsheets/blob/main/strings.pdf)

```{r}
#| label: a32
courses |> 
 mutate(year = str_c(str_sub(semester, 1, 4),
                      "-",
                     str_sub(semester, 5, 6))) |>
 select(semester, year) |>
  unique()
```

---

## Pivoting data

existing `faculty` data has one row per faculty, some with multiple departments (sometimes known as *wide* data)

```{r}
#| label: a33
#| echo: false
faculty
```

---

## Pivoting data

what if you instead want one row per faculty *per department*? (sometimes known as *long* data)

::: {style="font-size: 0.75em"}
```{r}
#| label: a34
#| echo: false
faculty |> 
  pivot_longer(dept1:dept2,
               names_to = "dept_no",
               values_to = "dept",
               values_drop_na = TRUE)
```
:::

---

## Pivoting data

the `pivot_longer` function lengthens data

```{r}
#| label: a35
faculty |> 
  pivot_longer(dept1:dept2,
               names_to = "dept_no",
               values_to = "dept",
              values_drop_na = TRUE) |>
  select(-year, -rank)
```

---

## Pivoting data

and `pivot_wider` does the opposite!

```{r}
#| label: a36
#| echo: false
courses |> head(2)
```

. . .

```{r}
#| label: a37
courses |> 
  pivot_wider(names_from = "level",
              values_from = "enrollment")
```

---

## Joining data

R has many useful functions for handling [relational data]{style="color: #cc0000"}

all you need is at least one [key variable]{style="color: #cc0000"} that connects data sets

`left_join` is most common, but there are [more](https://r4ds.had.co.nz/relational-data.html)

---

## Joining data

what's the average UG enrollment per year, per faculty rank?

. . .

::: {style="text-align: center"}
`faculty`
:::

::: {style="font-size: 0.45em"}
```{r}
#| label: a38
#| echo: false
reactable(faculty |> head(4))
```
:::

::: {style="text-align: center"}
`courses`
:::

::: {style="font-size: 0.45em"}
```{r}
#| label: a39
#| echo: false
reactable(courses |> head(4))
```
:::

. . .

::: {style="text-align: center"}
`faculty$id` is the same as `courses$faculty_id`
:::

---

## Joining data

<br>

what's the average [UG enrollment per year, per faculty]{style="color: #cc0000"} rank?

<br>

::: {style="font-size: 0.45em"}
```{r}
#| label: a40
#| echo: false
reactable(courses |> head(5))
```
:::

. . .

- filter to `UG` courses only
- create our `year` variable again 
- summarize `enrollment` by `year` and `faculty_id`

---

## Joining data

<br>

use the `<-` operator to create a new data frame `courses_UG`

```{r}
#| label: a41
#| eval: false
#| code-line-numbers: "1"
courses_UG <- courses |> 
 filter(level == "UG") |>
 mutate(year = str_c(str_sub(semester, 1, 4),
                      "-",
                      str_sub(semester, 5, 6)))
```

---

## Joining data

<br>

`filter` to undergraduate courses only and `mutate` a new academic year variable

```{r}
#| label: a42
#| eval: false
#| code-line-numbers: "2-5"
courses_UG <- courses |> 
 filter(level == "UG") |>
 mutate(year = str_c(str_sub(semester, 1, 4),
                      "-",
                      str_sub(semester, 5, 6)))
```

---

## Joining data

<br>

`group_by` year and faculty member; `summarize` enrollment

```{r}
#| label: a43
#| code-line-numbers: "6-7"
courses_UG <- courses |> 
 filter(level == "UG") |>
 mutate(year = str_c(str_sub(semester, 1, 4),
                      "-",
                     str_sub(semester, 5, 6))) |>
 group_by(year, faculty_id) |>
  summarize(enr = sum(enrollment))
```

. . .

<br>

::: {style="font-size: 0.45em"}
```{r}
#| label: a44
#| echo: false
reactable(courses_UG |> head(4))
```
:::

---

## Joining data

<br>

what's the average UG enrollment per year, per faculty rank?

::: columns
::: {.column width="60%"}
::: {style="text-align: center"}
`faculty`
:::
::: {style="font-size: 0.45em"}
```{r}
#| label: a45
#| echo: false
reactable(faculty)
```
:::
:::

::: {.column width="40%"}
::: {style="text-align: center"}
`courses_UG`
:::
::: {style="font-size: 0.45em"}
```{r}
#| label: a46
#| echo: false
reactable(courses_UG |> arrange(desc(year)))
```
:::
:::
:::

---

## Joining data

<br>

```{r}
#| label: a47
fac_enr <- faculty |>
  left_join(courses_UG, by = c("id" = "faculty_id",
                               "year" = "year"))
```

1. `fac_enr`: new data frame
2. `faculty`: data frame you're adding data to
3. `courses_UG`: data frame where the new data is coming from

. . .

::: {style="font-size: 0.45em"}
```{r}
#| label: a48
#| echo: false
reactable(fac_enr)
```
:::

---

## Joining data

<br>

what's the average UG enrollment per year, per faculty rank?

```{r}
#| label: a49
#| code-line-numbers: "4-5"
fac_enr <- faculty |> 
  left_join(courses_UG, by = c("id" = "faculty_id",
                              "year" = "year")) |>
 group_by(year, rank) |>
  summarize(avg_enr = mean(enr, na.rm = TRUE))
```

. . .

<br>

::: {style="font-size: 0.45em"}
```{r}
#| label: a50
#| echo: false
reactable(fac_enr |> arrange(desc(year)))
```
:::

# Data visualization

::: {.numbers}
[1 2 3 4]{style="opacity:0.25"} 5 [6]{style="opacity:0.25"}
:::

## ggplot2

<br>

`ggplot2` is the data visualization package that is loaded with the `tidyverse`

<br>

the [g]{style="color:#cc0000"}rammar of [g]{style="color:#cc0000"}raphics maps data to the aesthetic attributes of geometric points

. . .

<br>

encoding data into visual cues (e.g., length, color, position, size) is how we signify changes and comparisons

---

## Bar chart

```{r}
#| label: a51
#| code-line-numbers: "1-2|3|4"
#| eval: false
faculty |>
  count(rank) |>
  ggplot(aes(x = rank, y = n)) +
  geom_bar(stat = "identity")
```

. . .

to combine lines into one code chunk, use `+` instead of `|>`

. . .

```{r}
#| label: a52
#| echo: false
#| fig-height: 4
#| fig-align: "center"
faculty |>
  count(rank) |>
  ggplot(aes(x = rank, y = n)) +
  geom_bar(stat = "identity")
```

---

## Bar chart

can create a prettier plot pretty easily

```{r}
#| label: a53
#| code-fold: true
#| code-summary: "expand for full code"
#| fig-align: "center"
faculty |>
  count(rank) |>
  ggplot(aes(x = reorder(rank, -n), y = n)) +
  geom_bar(stat = "identity", fill = "#cc0000") +
  scale_y_continuous(expand = expansion(mult = c(0, 0.1))) +
  geom_text(aes(label = n), vjust = -0.5) +
  labs(x = NULL, y = NULL,
       title = "Count of faculty by rank, 2018-2021") +
  theme_linedraw() +
  theme(panel.grid.major.x = element_blank(),
        axis.ticks = element_blank())
```

---

## Line graph

<br>

```{r}
#| label: a54
#| code-line-numbers: "1-2|3|4"
#| eval: false
fac_enr |> 
 filter(!is.na(avg_enr)) |>
  ggplot(aes(x = year, y = avg_enr, group = rank, color = rank)) +
  geom_line()
```

. . .

```{r}
#| label: a55
#| echo: false
#| fig-height: 4
#| fig-align: "center"
fac_enr |> 
 filter(!is.na(avg_enr)) |>
  ggplot(aes(x = year, y = avg_enr, group = rank, color = rank)) +
  geom_line()
```

---

## Line graph

```{r}
#| label: a56
#| code-fold: true
#| code-summary: "expand for full code"
#| fig-align: "center"
fac_enr |> 
 filter(!is.na(avg_enr)) |>
  ggplot(aes(x = year, y = avg_enr, group = rank, color = rank)) +
  geom_line() +
  geom_point() +
  scale_color_brewer(type = "qual", palette = "Dark2") +
  labs(x = NULL, y = "Average enrollment",
       title = "Average undergraduate enrollment per rank over time") +
  theme_linedraw() +
  theme(panel.grid.major.x = element_blank(),
        axis.ticks = element_blank(),
        legend.title = element_blank(),
        legend.background = element_rect(fill = NA),
        legend.key = element_rect(fill = NA),
        legend.position = c(0.85, 0.82))
```

---

## ggplot2 resources

from [R for Data Science](https://r4ds.had.co.nz/data-visualisation.html)

<br>

[Data Visualization: a practical introduction](https://socviz.co/index.html#preface)

<br>

creating [custom themes](https://themockup.blog/posts/2020-12-26-creating-and-using-custom-ggplot2-themes/)

<br>

the [ggplot2 book](https://ggplot2-book.org/)

<br>

the [R graph gallery](https://r-graph-gallery.com/index.html)

---

## Putting it all together

<br>

with what we've done so far, your `.R` file could:

::: {.incremental}
- import your data files
- [document all data cleaning and preparation steps and decisions]{style="color:#cc0000"}
- produce a PPT-ready graphic summarizing your results
:::

. . .

::: {style="text-align: center"}
**and that file would make it extremely easy for you or someone else to reproduce this analysis with new data in six months**
:::

# Advanced topics

::: {.numbers}
[1 2 3 4 5]{style="opacity:0.25"} 6
:::

## R Markdown

<br>

using RStudio, create `.Rmd` documents that combine text, code, and graphics

. . .

<br>

many output formats: html, pdf, Word, slides

. . .

<br>

exceedingly useful for [parameterized reporting]{style="color:#cc0000"}: can create an R-based PDF report and generate it automatically for, say, each department

---

## Internal packages

<br>

you can also create your *own* packages!

. . .

<br>

your package can hold:

- common data sets that are used across projects
- custom `ggplot2` themes
- common functions and calculations (and their definitions!)

. . .

<br>

can be stored on a shared drive to facilitate collaboration

---

## R Markdown and package resources

<br>

**R Markdown**

the official [R Markdown website](https://rmarkdown.rstudio.com/lesson-1.html)

[R Markdown: The Definitive Guide](https://bookdown.org/yihui/rmarkdown/)

<br>

**internal packages**

a comprehensive [theoretical explainer](https://emilyriederer.netlify.app/post/team-of-packages/)

a [talk I gave](https://meghan.rbind.io/talk/boston-user/) earlier this year on the topic

# Learn more about R

## Resources

[R for Data Science](https://r4ds.had.co.nz/): the ultimate guide

<br>

[R for Excel users](https://rstudio-conf-2020.github.io/r-for-excel/): a very useful workshop

<br>

[STAT 545](https://stat545.com/): an online book on reproducible data analysis in R

<br>

the [RStudio Education](https://education.rstudio.com/) site

<br>

the [Learn `tidyverse`](https://www.tidyverse.org/learn/) site

<!--
vim: cole=0
-->
